xls/modules/zstd/README.md

```diff
@@ -0,0 +1,251 @@
+# ZSTD decoder
+
+The ZSTD decoder decompresses the correctly formed ZSTD frames and blocks.
+It implements the [RFC 8878](https://www.rfc-editor.org/rfc/rfc8878.html) decompression algorithm.
+Overview of the decoder architecture is presented on the diagram below.
+The decoder comprises:
+* frame decoder,
+* block dispatcher,
+* 3 types of processing units: RAW, RLE, and compressed,
+* command aggregator,
+* history buffer,
+* repacketizer.
+
+Incoming ZSTD frames are processed in the following order:
+1. magic number is detected,
+2. frame header is parsed,
+3. ZSTD data blocks are being redirected to correct processing unit based on the block header,
+4. processing unit results are aggregated in correct order into a stream
+and routed to the history buffer,
+5. data block outputs are assembled based on the history buffer contents and update history,
+6. decoded data is processed by repacketizer in order to prepare the final output of the decoder,
+7. (optional) calculated checksum is compared against frame checksum.
+
+![](img/ZSTD_decoder.png)
+
+## ZSTD decoder architecture
+
+### Top level Proc
+This state machine is responsible for receiving encoded ZSTD frames, buffering the input and passing it to decoder's internal components based on the state of the proc.
+The states defined for the processing of ZSTD frame are as follows:
```

Comment:
state diagram with https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams ?

Reply:
Added the state diagram of the Top Level Proc